### DISCUSSION TITLE
deb data-source handle extended name version (*.so version)

### DISCUSSION THREAD

**Discussion by ghost** [renovatebot/renovate#32704] [Request Help]
### How are you running Renovate?

Renovate 39.25.5 self hosted non-privileged via standard docker image.

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab 17.5.2 on-prem self hosted.

### Please tell us more about your question or problem

I'm using the [deb data-source](https://docs.renovatebot.com/modules/datasource/deb/#deb-datasource) and it works with custom regex for default package names and architectures.

Given own package list files:

file `apt.requirements.txt`:
```ini
liborsc-dev=11.1.0
liboiemodel-dev=5.0.0
liboieapp-dev=6.0.2
liboiealarm-dev=3.1.3
liboiediagnostics-dev=4.0.0
```
and file `apt.requirements_i386.txt`:

```ini
liboiersc:i386=11.1.0
liboieapp:i386=6.0.2
```
the custom regex parser

```json
    {
      "customType": "regex",
      "fileMatch": [
        "(^|/)apt\\.requirements(_.*)?\\.txt$"
      ],
      "datasourceTemplate": "deb",
      "matchStrings": [
        "(^|\\s)(?<depName>[a-zA-Z-_]+?)(:(?<depArch>.*))?=(?<currentValue>.*)"
      ],
      "registryUrlTemplate": "https://rtf.my.cloud/artifactory/oie-dev-public-debian?release=bookworm&components=main&binaryArch={{#if depArch }}{{depArch}}{{else}}amd64{{/if}}"
    }
```

Additionally we have some packages including the lib-so versions and this is not handled

file `apt.requirements.txt`:

```ini
liboiersc-common11.1=11.1.0
liboiersc11.1=11.1.0
liboiemodel5.0=5.0.0
liboieapp6=6.0.2
liboiealarm3.1=3.1.3
liboiediagnostics4.0=4.0.0
```

I see that [deb data source](https://github.com/renovatebot/renovate/commit/c3958c9bd6a675af8f763097cd66ce20fba23abf#diff-282f9822c61a9c5bdd21aab2b3c0a4831d1b41aa3a5bb86c28f811beb1e6483d) is handling the fields `Package` and `Version`.

How to add the lib-so version handling that must be included in the package name search, but return also new package name including the new lib-so version, e.g. `liboiersc-common11.1` -> `liboiersc-common12.0`  and update the line, e.g. `liboiersc-common11.1=11.1.0` -> `liboiersc-common12.0=12.0.0` ?

Any hints?

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by ghost** [comment]
@oxdev03 sorry to ping you directly here - as contributor of the deb [data-source](https://github.com/renovatebot/renovate/pull/30071) you may have some hints for me, thanks in advance.

**Reply by oxdev03** [reply]
Hi @andboss,
we didn't consider this usecase.
@secustor is this even possible to support this feature?
Since then it would be more a package "lookup" feature

**Reply by secustor** [reply]
What need to be added here?

**Reply by ghost** [reply]
IMHO: support using `packageNameTemplate` with pattern / regex possebility e.g. `{{{depName}}}*`or `{{{depName}}}\d+(\.\d+(\.\d+)?)?` to find the packages with lib-so version. Additionally provide the found package name to extract the lib-so version, like described [below](https://github.com/renovatebot/renovate/discussions/32704#discussioncomment-11381650) using `extractVersionTemplate`, but not clear if the package name can also be extracted with it?

**Reply by viceice** [reply]
I think it would be better to create a semi automatic preset with replacement rules.

**Reply by secustor** [reply]
The easiest fix for you would be to pull the versions out in variables, just as in the docs for the [custom manager with Dockerfiles](https://docs.renovatebot.com/modules/manager/regex/#regular-expression-capture-groups). 

Outside of that you have to go with dynamically generated / auto updated replacement rules as `viceice` suggested.

**Reply by ghost** [reply]
Thanks, i don't get it. The example below shows that i've already extracting the lib-so versions. BUT the data source is searching for exact match of the package name and therefore cannot find the packages with the lib-so versions(s) attached.

What are dynamically generated replacements? I'm investigating. but cannot find an clear example. Any example would be appreciated...

**Reply by secustor** [reply]
What I have meant with dynamic replacements is you can let Renovate update your replacements rules via custom manger in your `renovate.json` or other config files.

**Comment by ghost** [comment]
P.S.: previously i used custom-datasource that used artifactory search functionality to find the packages via `{{{depName}}}*{{{depArch}}}` in a `packageNameTemplate` that returned the **filename** and we extracted the `depSoVerNew` returned via `extractVersionTemplate` and applied back to the listing file with `autoReplaceStringTemplate`. That was somehow difficult to understand ;) 

```json
   {
     "customType": "regex",
      "fileMatch": [
        "(^|/)apt\\.requirements(_.*)?\\.apt\\.txt$"
      ],
     "matchStrings": [
       "(^|\\s)(?<depName>[a-zA-Z-_]+?)(?<depSoVerCurrent>\\d+(\\.\\d+(\\.\\d+)?)?)?(?<depArchFull>:(?<depArch>.*))?=(?<currentValue>.*)"
     ],
     "extractVersionTemplate": "^.*/{{{depName}}}(?<depSoVerNew>[\\d\\.]+)_(?<version>.*?)_{{{depArch}}}.*$",
     "datasourceTemplate": "artifactory-search",
     "registryUrlTemplate": "https://rtf.my.cloud/artifactory",
     "packageNameTemplate": "repos=oie-dev-public-debian&name={{{depName}}}*{{{depArch}}}",
     "autoReplaceStringTemplate": "{{{depName}}}{{{depSoVerNew}}}{{{depArchFull}}}={{{newValue}}}"
   },
```

The `Packages` file fetched by the deb data-source is including all this, e.g.:
```
Package: liboiersc-common12.0
Source: rsc
Version: 12.0.0
Architecture: all
Filename: pool/main/r/rsc/liboiersc-common12.0_12.0.0_all.deb
```

If we could have the `Filename` propagated as variable somehow and have the Package searched by pattern `package*` or regex: `package\d+(\.\d+(\.\d+)?)?` or just the new `Package` name found it may help. WDYT?
